src/hypothesis/strategies.py

```diff
@@ -775,19 +778,126 @@ def seed_random(seed):
 @cacheable
 @defines_strategy
 def builds(target, *args, **kwargs):
-    """Generates values by drawing from args and kwargs and passing them to
-    target in the appropriate argument position.
+    """Generates values by drawing from ``args`` and ``kwargs`` and passing
+    them to ``target`` in the appropriate argument position.
 
-    e.g. builds(target, integers(), flag=booleans()) would draw an
-    integer i and a boolean b and call target(i, flag=b).
+    e.g. ``builds(target, integers(), flag=booleans())`` would draw an
+    integer ``i`` and a boolean ``b`` and call ``target(i, flag=b)``.
+
+    If ``target`` has type annotations, they will be used to infer a strategy
+    for required arguments that were not passed to builds.  You can also tell
+    builds to infer a strategy for an optional argument by passing the special
+    value :const:`hypothesis.infer` as a keyword argument to
+    builds, instead of a strategy for that argument to ``target``.
 
     """
+    from hypothesis.searchstrategy import types
+
+    if infer in args:
+        # Avoid an implementation nightmare juggling tuples and worse things
+        raise InvalidArgument('infer was passed as a positional argument to '
+                              'builds(), but is only allowed as a keyword arg')
+    hints = get_type_hints(target)
+    for kw in [k for k, v in kwargs.items() if v is infer]:
+        if kw not in hints:
+            raise InvalidArgument(
+                'passed %s=infer for %s, but %s has no type annotation'
+                % (kw, target.__name__, kw))
+        kwargs[kw] = from_type(hints[kw])
+
+    try:
+        spec = getfullargspec(target)
+    except TypeError:
+        pass
+    else:
+        for ms in set(hints) & types.required_args(spec, args, kwargs):
+            kwargs[ms] = from_type(hints[ms])
+
     return tuples(tuples(*args), fixed_dictionaries(kwargs)).map(
         lambda value: target(*value[0], **value[1])
     )
 
 
 @cacheable
+def from_type(thing):
+    """Looks up the appropriate search strategy for the given type.
+
+    ``from_type`` is used internally to fill in missing arguments to
+    :func:`~hypothesis.strategies.builds` and can be used interactively
+    to explore what strategies are available or to debug type resolution.
+
+    You can use :func:`~hypothesis.strategies.register_type_strategy` to
+    handle your custom types, or to globally redefine certain strategies -
+    for example excluding NaN from floats, or use timezone-aware instead of
+    naive time and datetime strategies.
+
+    The resolution tries four options, in this order:
```

Comment:
I think I'd like to caveat this with saying that although this is documented it is liable to change in minor releases. I suspect when this makes it into the wild we're going to find a lot of cases where people are not happy with the resolved strategy for a type and we're going to want to tinker with the logic.